File: 2024/python/9/part1.py
raw = open('input.dat').readlines()
raw = [line.strip() for line in raw]
raw = raw[0]

if False:
    raw = '2333133121414131402'

cnt = 0
block = []
for i, num in enumerate(raw):
    num = int(num)
    if i % 2 == 0:
        block += [str(cnt)] * num
        cnt += 1
    else:
        block += ['.' for _ in range(num)]

i = 0
checksum = 0
while i < len(block):
    while block[i] == '.':
        block[i] = block.pop()
    checksum += int(block[i]) * i
    i += 1

print(checksum)

File: 2024/go/3/part1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
)

func main() {
	file, err := os.Open("input.dat")
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	reg := regexp.MustCompile(`mul\((\d+),(\d+)\)`)
	res := 0

	for scanner.Scan() {
		line := scanner.Text()
		matches := reg.FindAllStringSubmatch(line, -1)

		for _, match := range matches {
			res += toInt(match[1]) * toInt(match[2])
		}
	}
	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	fmt.Println(res)
}

func toInt(s string) int {
	num, err := strconv.Atoi(s)
	if err != nil {
		fmt.Printf("Error converting '%s' to int: %v\n", s, err)
		return 0
	}
	return num
}
